vnext/ReactWindowsCore/WinRTWebSocketResource.cpp

```diff
@@ -0,0 +1,363 @@
+    }
+
+    string message;
+    bool isBinary;
+    size_t length;
+    std::tie(message, isBinary) = m_writeQueue.front();
+    m_writeQueue.pop();
+
+    if (isBinary)
+    {
+      m_socket.Control().MessageType(SocketMessageType::Binary);
+
+      auto buffer = CryptographicBuffer::DecodeFromBase64String(Utf8ToUtf16(std::move(message)));
+      length = buffer.Length();
+      m_writer.WriteBuffer(buffer);
+    }
+    else
+    {
+      m_socket.Control().MessageType(SocketMessageType::Utf8);
+
+      length = message.size();
+      winrt::array_view<const uint8_t> arr(
+        CheckedReinterpretCast<const uint8_t*>(message.c_str()),
+        CheckedReinterpretCast<const uint8_t*>(message.c_str()) + message.length());
+      m_writer.WriteBytes(arr);
+    }
+
+    co_await m_writer.StoreAsync();
+
+    if (m_writeHandler)
+    {
+      m_writeHandler(length);
+    }
+
+    if (!m_writeQueue.empty())
+    {
+      PerformWrite();
+    }
+  }
+  catch (hresult_error const& e)
+  {
+    hr = e.code();
+  }
+
+  if (!SUCCEEDED(hr))
+  {
+    if (m_errorHandler)
+    {
+      m_errorHandler({ Utf16ToUtf8(hresult_error(hr).message()), ErrorType::Ping });
+    }
+  }
+}
+
+fire_and_forget WinRTWebSocketResource::PerformClose()
+{
+  co_await resume_background();
+
+  m_socket.Close(static_cast<uint16_t>(m_closeCode), Utf8ToUtf16(m_closeReason));
+
+  co_return;
+}
+
+void WinRTWebSocketResource::OnMessageReceived(IWebSocket const& sender, MessageWebSocketMessageReceivedEventArgs const& args)
+{
+  try
+  {
+    string response;
+    DataReader reader = args.GetDataReader();
+    auto len = reader.UnconsumedBufferLength();
+    if (args.MessageType() == SocketMessageType::Utf8)
+    {
+      reader.UnicodeEncoding(UnicodeEncoding::Utf8);
+      vector<uint8_t> data(len);
+      reader.ReadBytes(data);
+
+      response = string(CheckedReinterpretCast<char*>(data.data()), data.size());
+    }
+    else
+    {
+      auto buffer = reader.ReadBuffer(len);
+      winrt::hstring data = CryptographicBuffer::EncodeToBase64String(buffer);
+
+      response = Utf16ToUtf8(std::wstring_view(data));
+    }
+
+    if (m_readHandler)
+      m_readHandler(response.length(), response);//TODO: move?
+  }
+  catch (hresult_error const& e)
+  {
+    //TODO: Re-enable (fix "OutputDebugStringW': function does not take 2 arguments")
+    //OutputDebugString("Read failed (0x%8X) %ls\n", e);
+    m_errorHandler({ Utf16ToUtf8(e.message()), ErrorType::Receive });
+  }
+}
+
+void WinRTWebSocketResource::OnClosed(IWebSocket const& sender, WebSocketClosedEventArgs const& args)
+{
+  if (m_closeHandler)
+  {
+    //TODO: Parameterize (pass via member variables?)
+    m_closeHandler(CloseCode::Normal, "Closing");
+  }
+}
+
+void WinRTWebSocketResource::Stop()
+{
+  // Ensure sequence of other operations
+  if (m_connectRequested)
+  {
+    // m_connectPerformed.Wait();//TODO: Set timeout?
+    m_connectPerformed.get_future().wait();
```

Comment:
Would be great if we could do this async.

Reply:
It's mostly been replaced by `co_await`.

May still need it for the final synchronization at object's destruction.
This will be ruled by the integration tests passes.